extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java

```diff
@@ -363,30 +366,41 @@ public void run()
           // The retrying business is because the KafkaConsumer throws OffsetOutOfRangeException if the seeked-to
           // offset is not present in the topic-partition. This can happen if we're asking a task to read from data
           // that has not been written yet (which is totally legitimate). So let's wait for it to show up.
-          final ConsumerRecords<byte[], byte[]> records = RetryUtils.retry(
-              new Callable<ConsumerRecords<byte[], byte[]>>()
-              {
-                @Override
-                public ConsumerRecords<byte[], byte[]> call() throws Exception
+          ConsumerRecords<byte[], byte[]> records = ConsumerRecords.empty();
+          try {
+            polling = true;
+            records = RetryUtils.retry(
+                new Callable<ConsumerRecords<byte[], byte[]>>()
                 {
-                  try {
+                  @Override
+                  public ConsumerRecords<byte[], byte[]> call() throws Exception
+                  {
                     return consumer.poll(POLL_TIMEOUT);
                   }
-                  finally {
-                    status = Status.READING;
-                  }
-                }
-              },
-              new Predicate<Throwable>()
-              {
-                @Override
-                public boolean apply(Throwable input)
+                },
+                new Predicate<Throwable>()
                 {
-                  return input instanceof OffsetOutOfRangeException;
-                }
-              },
-              Integer.MAX_VALUE
-          );
+                  @Override
+                  public boolean apply(Throwable input)
+                  {
+                    return input instanceof OffsetOutOfRangeException;
+                  }
+                },
+                Integer.MAX_VALUE
+            );
+            synchronized (pollingSyncLock) {
+              if (Thread.interrupted()) {
+                throw new InterruptedException();
+              }
+              polling = false;
```

Comment:
The logic here is that if the thread was interrupted, it could be a) because of a pause command breaking out of the retry or b) as part of the shutdown process. If it's part of the pause, `polling` will have been set to `false` by the pause method before interrupting the thread, otherwise `polling` will be `true`. If we set `polling = false` first before throwing the interrupt, then when it is caught on line 398 it will always be treated as a non-shutdown interrupt, the interrupt flag won't be re-set, and the thread won't die.